featurebyte/models/precomputed_lookup_feature_table.py

```diff
@@ -0,0 +1,294 @@
+"""
+Classes to support precomputed lookup feature tables
+"""
+
+from typing import Dict, List, Optional, Tuple, cast
+
+import hashlib
+import json
+
+from sqlglot import expressions
+from sqlglot.expressions import Expression, Select
+
+from featurebyte.enum import SpecialColumnName
+from featurebyte.models.base import PydanticObjectId
+from featurebyte.models.entity_lookup_feature_table import get_entity_lookup_graph
+from featurebyte.models.entity_universe import (
+    CURRENT_FEATURE_TIMESTAMP_PLACEHOLDER,
+    EntityUniverseModel,
+    EntityUniverseParams,
+    get_combined_universe,
+)
+from featurebyte.models.feature_list import FeatureListModel
+from featurebyte.models.feature_store import FeatureStoreModel
+from featurebyte.models.offline_store_feature_table import (
+    OfflineStoreFeatureTableModel,
+    PrecomputedLookupFeatureTableInfo,
+)
+from featurebyte.models.parent_serving import EntityLookupStep
+from featurebyte.models.sqlglot_expression import SqlglotExpressionModel
+from featurebyte.query_graph.model.entity_lookup_plan import EntityLookupPlanner
+from featurebyte.query_graph.model.entity_relationship_info import (
+    EntityAncestorDescendantMapper,
+    EntityRelationshipInfo,
+)
+from featurebyte.query_graph.sql.common import construct_cte_sql, quoted_identifier
+from featurebyte.query_graph.sql.parent_serving import construct_request_table_with_parent_entities
+
+
+def get_lookup_steps_unique_identifier(lookup_steps: List[EntityRelationshipInfo]) -> str:
+    """
+    Get a short unique identifier for a list of lookup steps
+
+    Parameters
+    ----------
+    lookup_steps: List[EntityRelationshipInfo]
+        Relationships used for parent entity lookup
+
+    Returns
+    -------
+    str
+    """
+    hasher = hashlib.shake_128()
+    for lookup_step in lookup_steps:
+        hasher.update(json.dumps(lookup_step.json_dict()).encode("utf-8"))
+    return hasher.hexdigest(3)
+
+
+def get_precomputed_lookup_feature_table_name(
+    source_feature_table_name: str,
+    serving_names: List[str],
+    lookup_steps: List[EntityRelationshipInfo],
+) -> str:
+    """
+    Construct the name of a precomputed lookup feature table
+
+    Parameters
+    ----------
+    source_feature_table_name: str
+        Name of the offline store feature table that this table originates from
+    serving_names: List[str]
+        Serving names of the table
+    lookup_steps: List[EntityRelationshipInfo]
+        Relationships used for parent entity lookup
+
+    Returns
+    -------
+    str
+    """
+    serving_names_suffix = OfflineStoreFeatureTableModel.get_serving_names_for_table_name(
+        serving_names
+    )
+    unique_identifier = get_lookup_steps_unique_identifier(lookup_steps)
+    return f"{source_feature_table_name}_via_{serving_names_suffix}_{unique_identifier}"
```

Comment:
Max table name length for MySQL is 64 (for DataBricks is 255). Do we need to worry about exceeding table name limit here?

Reply:
That might occur since `source_feature_table_name` can reach that limit, though there should be plenty of room for databricks. I suggest we revisit the limits again later to account for the new suffixes.